src/sota_tools/ostree_object.cc

```diff
@@ -235,15 +236,37 @@ void OSTreeObject::Upload(const TreehubServer &push_target, CURLM *curl_multi_ha
   request_start_time_ = std::chrono::steady_clock::now();
 }
 
+void OSTreeObject::PresenceUnknown(RequestPool &pool, const long rescode) {
+  is_on_server_ = PresenceOnServer::kObjectStateUnknown;
+  LOG_WARNING << "OSTree query reported an error code: " << rescode << " retrying...";
+  LOG_DEBUG << "Http response code:" << rescode;
+  LOG_DEBUG << http_response_.str();
+  last_operation_result_ = ServerResponse::kTemporaryFailure;
+  pool.AddQuery(this);
+}
+
+void OSTreeObject::ObjectMissing(RequestPool &pool, const long rescode) {
+  LOG_WARNING << "OSTree upload reported an error code:" << rescode << " retrying...";
+  LOG_DEBUG << "Http response code:" << rescode;
+  LOG_DEBUG << http_response_.str();
+  is_on_server_ = PresenceOnServer::kObjectMissing;
+  last_operation_result_ = ServerResponse::kTemporaryFailure;
+  pool.AddUpload(this);
+}
+
 void OSTreeObject::CurlDone(CURLM *curl_multi_handle, RequestPool &pool) {
   refcount_--;            // Because curl now doesn't have a reference to us
   assert(refcount_ > 0);  // At least our parent should have a reference to us
 
+  char *url = NULL;
+  curl_easy_getinfo(curl_handle_, CURLINFO_EFFECTIVE_URL, &url);
   long rescode = 0;  // NOLINT
   curl_easy_getinfo(curl_handle_, CURLINFO_RESPONSE_CODE, &rescode);
   if (current_operation_ == CurrentOp::kOstreeObjectPresenceCheck) {
-    if (rescode == 200) {
-      LOG_INFO << "Already Present:" << object_name_;
+    if (!url || !strstr(url, object_name_.c_str())) {
```

Comment:
You're trying to see if the server sent us a redirect?

Reply:
Nah, just verifying that the object the server found matches the object we have locally.